storage/src/robocorp/storage/__init__.py

```diff
@@ -22,142 +22,128 @@ def list_assets() -> List[AssetMeta]:
-        wait: Blocks until the new value is reflected within the asset.
+        name: Name of the existing or new asset to create (if missing)
+        value: The new value set within the asset
+        wait: Wait for value to be set succesfully
 
     Raises:
-        AssetNotFound: When the queried asset doesn't exist.
+        RuntimeError: Asset upload failed
     """
-    existing_asset = _get_asset(name, raise_if_missing=False)
-    if existing_asset:
-        asset_id = existing_asset["id"]
-        LOGGER.debug("Updating already existing asset with ID %r.", asset_id)
-    else:
+    try:
+        asset_id = _get_asset(name)["id"]
+        LOGGER.debug("Updating existing asset with id: %r", asset_id)
+    except AssetNotFound:
         asset_id = _create_asset(name)["id"]
-        LOGGER.debug("Updating newly created asset with ID %r.", asset_id)
+        LOGGER.debug("Created new asset with id: %r", asset_id)
 
     assets_client = _get_assets_client()
     body = {"content_type": "text/plain"}
     upload_data = assets_client.post(f"{asset_id}/upload", json=body).json()
     assets_client.put(upload_data["upload_url"], data=value, headers={})
 
-    if wait:
-        LOGGER.info(
-            "Waiting for the sent value to be set successfully in the %r asset...", name
-        )
-        upload_url = f"{asset_id}/uploads/{upload_data['id']}"
-        while True:
-            upload_data = assets_client.get(upload_url).json()
-            status = upload_data["status"]
-            if status == "pending":
-                sleep_time = round(random.uniform(0, 1), 2)
-                LOGGER.debug("Asset upload still pending, sleeping %.2f...", sleep_time)
-                time.sleep(sleep_time)
-                continue
-            elif status == "completed":
-                break
-            elif status == "failed":
-                raise ValueError(
-                    f"asset {name!r} upload failed with {upload_data['reason']!r}"
-                )
-            else:
-                raise TypeError(f"asset {name!r} got invalid status {status!r}")
-
-    LOGGER.info("Asset with name %r set successfully.", name)
+    if not wait:
+        return
+
+    LOGGER.info("Waiting for asset %r value to update succesfully", name)
+    upload_url = f"{asset_id}/uploads/{upload_data['id']}"
+    while True:
+        upload_data = assets_client.get(upload_url).json()
+        status = upload_data["status"]
+        if status == "pending":
+            sleep_time = round(random.uniform(0, 1), 2)
+            LOGGER.debug("Asset upload still pending, sleeping %.2f...", sleep_time)
+            time.sleep(sleep_time)
+            continue
+        elif status == "completed":
+            break
+        elif status == "failed":
+            reason = upload_data["reason"]
+            raise RuntimeError(f"Asset {name!r} upload failed: {reason!r}")
+        else:
+            raise RuntimeError(f"Asset {name!r} got invalid status: {status!r}")
```

Comment:
Maybe is even better to have a custom error called `AssetUploadError` deriving from `RuntimeError` (or a better exception class).

Reply:
I think RuntimeError (as a base class) is a valid choice since they're not dictated by the user input, but something unexpected from the server. We could make a custom one though, although we'd probably want to bubble that up from all API call errors.